File: scripts/plot_damage_fractional.py
#!/usr/bin/env python3
"""
plot_damage_fractional.py

Plot fractional-position aDNA damage profiles from pre-computed profiles
output by aggregate_sample.py (*_damage_fractional_profile.tsv files).

X axis: fractional position along the read (0.0 = 5' end, 1.0 = 3' end)
Y axis: fraction of unclassified k-mers (%)

Stratifying by read length exposes:
  - Genuine terminal damage spikes at both ends (all strata)
  - The read-length artifact: for short reads, opposite-end damage zones
    overlap, producing a U-shaped profile with no flat interior
  - The plateau region is only interpretable for longer read strata

Multiple samples (e.g. damaged vs undamaged) can be overlaid per stratum
by passing multiple --fractional files.

Usage:
  # Single sample, select by taxid
  python plot_damage_fractional.py \\
    --fractional  sample_fractional_profile.tsv \\
    --taxid       1649845 \\
    --output      damage_fractional.pdf

  # Single sample, select by species name (from --species-map aggregation)
  python plot_damage_fractional.py \\
    --fractional  sample_fractional_profile.tsv \\
    --species     "Streptococcus mutans" \\
    --output      damage_fractional.pdf

  # Two samples overlaid
  python plot_damage_fractional.py \\
    --fractional  dam1_fractional_profile.tsv dam0_fractional_profile.tsv \\
    --label       "aDNA damage"  "No damage" \\
    --taxid       1649845 \\
    --output      damage_fractional_compare.pdf \\
    [--strata "31-40" "41-55" "56-75" "76-100"]
"""

import argparse
import sys

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
import numpy as np
import pandas as pd

from hit_species_selection import select_hit_species


# ---------------------------------------------------------------------------
# Defaults
# ---------------------------------------------------------------------------

DEFAULT_STRATA = [
    (31,  40,  "#d73027", "31–40 bp"),
    (41,  55,  "#fc8d59", "41–55 bp"),
    (56,  75,  "#4575b4", "56–75 bp"),
    (76, 100,  "#313695", "76–100 bp"),
]
DEFAULT_SAMPLE_COLORS = ["#d6604d", "#2166ac", "#4dac26", "#8073ac"]
DEFAULT_SAMPLE_STYLES = ["-", "--", "-.", ":"]


def resolve_keys(global_path: str | None, pvalue_threshold: float = 0.05) -> list:
    """
    Auto-select taxa to plot from a damage_global.tsv file.
    Prefers taxa with damage_pvalue < threshold, falls back to top 3 by score.
    Returns a list of species_name strings (or taxids if species_name absent).
    """
    if global_path is None:
        return []
    df = pd.read_csv(global_path, sep="\t")
    if df.empty:
        return []

    group_col = "species_name" if "species_name" in df.columns else "taxid"

    if "damage_pvalue" in df.columns:
        sig_mask = df["damage_score"].gt(0) & df["damage_pvalue"].lt(pvalue_threshold)
        sig_keys = set(df.loc[sig_mask, group_col].dropna())
        if sig_keys:
            keys = (
                df[df[group_col].isin(sig_keys)]
                .groupby(group_col)["damage_pvalue"]
                .mean().dropna()
                .sort_values(ascending=True)
                .index.tolist()
            )
            print(
                f"Auto-selected {len(keys)} taxa with damage_pvalue < {pvalue_threshold}.",
                file=sys.stderr,
            )
            return keys
        print(
            f"No taxa pass p < {pvalue_threshold}. Falling back to top 3 by score.",
            file=sys.stderr,
        )

    return (
        df.groupby(group_col)["damage_score"]
        .mean().dropna()
        .sort_values(ascending=False)
        .head(3).index.tolist()
    )


def parse_strata_spec(specs: list[str]) -> list[tuple[int, int, str, str]]:
    """Parse 'lo-hi' strings into strata tuples, reusing default colors."""
    result = []
    for i, spec in enumerate(specs):
        lo, hi = (int(x) for x in spec.split("-"))
        color = DEFAULT_STRATA[i][2] if i < len(DEFAULT_STRATA) else "#888888"
        result.append((lo, hi, color, f"{lo}–{hi} bp"))
    return result


# ---------------------------------------------------------------------------
# Load pre-computed fractional profiles
# ---------------------------------------------------------------------------

def load_fractional_table(path: str) -> pd.DataFrame:
    """
    Load a *_fractional_profile.tsv with only columns needed for plotting.
    """
    wanted = {
        "taxid",
        "species_name",
        "stratum",
        "bin",
        "n_reads",
        "n_total",
        "n_unclassified",
    }
    return pd.read_csv(path, sep="\t", usecols=lambda c: c in wanted)


def build_key_lookup(df: pd.DataFrame) -> dict[str, dict]:
    """
    Build key -> row-index lookups so per-key plotting doesn't re-scan the table.
    """
    species_lookup: dict[str, pd.Index] = {}
    taxid_lookup: dict[int, pd.Index] = {}

    if "species_name" in df.columns:
        sp = df["species_name"].fillna("").astype(str).str.strip()
        mask = sp != ""
        if mask.any():
            species_lookup = {
                str(k): idx
                for k, idx in df.loc[mask].groupby(sp[mask], sort=False).groups.items()
            }

    if "taxid" in df.columns:
        tax = pd.to_numeric(df["taxid"], errors="coerce")
        mask = tax.notna()
        if mask.any():
            tax_i = tax[mask].astype(int)
            taxid_lookup = {
                int(k): idx
                for k, idx in df.loc[mask].groupby(tax_i, sort=False).groups.items()
            }

    return {"species": species_lookup, "taxid": taxid_lookup}


def subset_for_key(df: pd.DataFrame, lookup: dict[str, dict], key) -> pd.DataFrame:
    """Return subset rows for a key using precomputed lookup."""
    if isinstance(key, int):
        idx = lookup["taxid"].get(int(key))
    else:
        idx = lookup["species"].get(str(key))
    if idx is None:
        return df.iloc[0:0]
    return df.loc[idx]


def filter_missing_keys(keys, lookups):
    """Drop keys that are absent from all loaded fractional tables."""
    if not keys:
        return keys
    if isinstance(keys[0], int):
        available = set()
        for lookup in lookups:
            available.update(lookup["taxid"].keys())
    else:
        available = set()
        for lookup in lookups:
            available.update(lookup["species"].keys())
    kept = [k for k in keys if k in available]
    dropped = len(keys) - len(kept)
    if dropped > 0:
        print(
            f"WARNING: dropped {dropped} selected taxa with no fractional rows.",
            file=sys.stderr,
        )
    return kept


def build_arrays(
    df: pd.DataFrame,
    strata: list[tuple[int, int, str, str]],
    n_bins: int,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Reconstruct (frac_total, frac_unc, read_counts) arrays from a filtered
    fractional profile dataframe.

    Returns:
      frac_total  shape (n_strata, n_bins)
      frac_unc    shape (n_strata, n_bins)
      read_counts shape (n_strata,)
    """
    ns = len(strata)
    frac_total  = np.zeros((ns, n_bins), dtype=np.int64)
    frac_unc    = np.zeros((ns, n_bins), dtype=np.int64)
    read_counts = np.zeros(ns, dtype=np.int64)

    for s_idx, (lo, hi, _, _) in enumerate(strata):
        sub = df[df["stratum"] == f"{lo}-{hi}"]
        if sub.empty:
            continue
        read_counts[s_idx] = int(sub["n_reads"].iloc[0])
        bins  = sub["bin"].values.astype(int)
        valid = (bins >= 0) & (bins < n_bins)
        frac_total[s_idx, bins[valid]] = sub["n_total"].values[valid].astype(np.int64)
        frac_unc[s_idx,   bins[valid]] = sub["n_unclassified"].values[valid].astype(np.int64)

    return frac_total, frac_unc, read_counts


# ---------------------------------------------------------------------------
# Plotting
# ---------------------------------------------------------------------------

def make_plot(
    all_totals, all_uncs, all_read_counts,
    labels, sample_colors, sample_styles,
    strata, n_bins, key, output_path=None,
):
    n_strata = len(strata)
    bin_centers = (np.arange(n_bins) + 0.5) / n_bins

    ncols = 2
    nrows = (n_strata + 1) // 2
    fig, axes = plt.subplots(nrows, ncols, figsize=(11, 4 * nrows), squeeze=False)
    fig.suptitle(
        f"aDNA damage profiles — fractional position by read length\n"
        f"{key}  |  "
        + ("  ".join(f"{lbl}: solid" if i == 0 else f"{lbl}: dashed"
                     for i, lbl in enumerate(labels))
           if len(labels) > 1 else labels[0]),
        fontsize=10,
    )

    for s_idx, (lo, hi, s_color, s_label) in enumerate(strata):
        row, col = divmod(s_idx, ncols)
        ax = axes[row][col]

        for totals, uncs, read_counts, label, s_col, s_sty in zip(
            all_totals, all_uncs, all_read_counts,
            labels, sample_colors, sample_styles,
        ):
            t = totals[s_idx].astype(float)
            u = uncs[s_idx].astype(float)
            mask = t > 0
            frac = np.where(mask, u / np.where(mask, t, 1) * 100, np.nan)
            n_reads = int(read_counts[s_idx])
            lbl = f"{label}  (n={n_reads:,})"
            ax.plot(bin_centers[mask], frac[mask],
                    color=s_col, ls=s_sty, lw=2, alpha=0.9, label=lbl)

        ax.axvline(0.5, color="gray", lw=0.8, ls=":", alpha=0.4, label="Midpoint")
        ax.set_title(s_label, fontsize=9)
        ax.set_xlabel("Fractional position (0 = 5′, 1 = 3′)", fontsize=8)
        ax.set_ylabel("Unclassified k-mers (%)", fontsize=8)
        ax.set_xlim(0, 1)
        ax.set_ylim(bottom=0)
        ax.legend(fontsize=7.5, frameon=False)
        ax.spines[["top", "right"]].set_visible(False)
        ax.tick_params(labelsize=7)

    for s_idx in range(n_strata, nrows * ncols):
        row, col = divmod(s_idx, ncols)
        axes[row][col].set_visible(False)

    fig.text(
        0.5, -0.01,
        "Short reads (31-40 bp): damage zones overlap across entire read, "
        "no flat interior. Longer reads show a flat central plateau.",
        ha="center", fontsize=8, color="gray", style="italic",
    )
    fig.tight_layout()
    if output_path is not None:
        fig.savefig(output_path, bbox_inches="tight", dpi=150)
        print(f"Saved {output_path}", file=sys.stderr)
    return fig


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def parse_args():
    p = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    p.add_argument("--fractional", nargs="+", required=True,
                   help="*_damage_fractional_profile.tsv file(s) from aggregate_sample.py")
    p.add_argument("--output", required=True,
                   help="Output plot path (.pdf or .png)")
    p.add_argument("--label",  nargs="+", default=[],
                   help="Sample labels (one per --fractional file)")
    p.add_argument("--strata", nargs="+", default=[],
                   help="Read-length strata to plot as 'lo-hi' (default: all four default strata)")

    key_grp = p.add_mutually_exclusive_group(required=False)
    key_grp.add_argument("--taxid",   type=int,
                         help="Integer taxid to plot (for per-taxid profiles)")
    key_grp.add_argument("--species", type=str,
                         help="Species name to plot (for species-aggregated profiles)")

    p.add_argument("--global", dest="global_tsv", default=None,
                   help="damage_global.tsv for auto-selecting significant taxa "
                        "(used when --taxid/--species are omitted and --hits not provided)")
    p.add_argument("--hits",       default=None,
                   help="Integrated summary table (.tsv/.tsv.gz); use sample hit species as key list")
    p.add_argument("--sample-id",  default=None,
                   help="Sample ID to filter from --hits table")
    p.add_argument("--hits-required-flags", nargs="+", default=[],
                   help="Required tokens in hit_criteria_flag for --hits selection "
                        "(all must be present)")
    p.add_argument("--max-keys", type=int, default=200,
                   help="Maximum number of taxa/pages to plot for auto or --hits selection (default: 200; <=0 disables cap)")
    p.add_argument("--pvalue-threshold", type=float, default=0.05,
                   help="Max damage_pvalue for auto-selection (default: 0.05)")

    return p.parse_args()


def _resolve_required_hit_flags(args: argparse.Namespace) -> list[str]:
    if args.hits_required_flags:
        return args.hits_required_flags
    return ["damage_pvalue", "within_genus_relative_abundance", "classified_rate"]


def write_empty_plot(output_path: str, note: str) -> None:
    fig, ax = plt.subplots(figsize=(9, 4.5))
    ax.axis("off")
    ax.text(
        0.5, 0.62,
        "No fractional damage plots were generated",
        ha="center", va="center", fontsize=13, fontweight="bold",
    )
    ax.text(
        0.5, 0.42,
        note,
        ha="center", va="center", fontsize=9.5,
    )
    fig.tight_layout()
    fig.savefig(output_path, bbox_inches="tight", dpi=150)
    plt.close(fig)
    print(f"Saved {output_path} (empty fractional)", file=sys.stderr)


def main():
    args = parse_args()

    strata = parse_strata_spec(args.strata) if args.strata else DEFAULT_STRATA

    # Resolve which taxa to plot: --hits takes priority, then explicit args, then auto
    hit_species: list[str] = []
    required_hit_flags = _resolve_required_hit_flags(args)
    use_hits = bool(args.hits and args.sample_id)
    if use_hits:
        try:
            hit_species, hit_info = select_hit_species(
                hits_path=args.hits,
                sample_id=str(args.sample_id),
                required_flag_tokens=required_hit_flags,
                max_keys=args.max_keys,
            )
            if not hit_species:
                print(
                    "WARNING: no hit species matched all required hit flags for this sample.",
                    file=sys.stderr,
                )
            if hit_info["n_truncated"] > 0:
                print(
                    f"WARNING: selected hit species truncated to {len(hit_species)} "
                    f"(dropped {hit_info['n_truncated']} by --max-keys).",
                    file=sys.stderr,
                )
        except (FileNotFoundError, pd.errors.EmptyDataError, pd.errors.ParserError, EOFError, OSError) as e:
            print(f"WARNING: could not read --hits file ({args.hits}): {e}", file=sys.stderr)
            pass

    if use_hits:
        keys = hit_species
    elif args.taxid is not None:
        keys = [args.taxid]
    elif args.species is not None:
        keys = [args.species]
    else:
        keys = resolve_keys(args.global_tsv, args.pvalue_threshold)
        if args.max_keys > 0 and len(keys) > args.max_keys:
            dropped = len(keys) - args.max_keys
            keys = keys[: args.max_keys]
            print(
                f"WARNING: auto-selected taxa truncated to {len(keys)} "
                f"(dropped {dropped} by --max-keys).",
                file=sys.stderr,
            )

    if not keys:
        print(
            "No taxa to plot. Provide --taxid, --species, or --global with "
            "significant taxa.",
            file=sys.stderr,
        )
        write_empty_plot(
            args.output,
            "No taxa were selected from --hits, explicit args, or --global fallback.",
        )
        sys.exit(0)

    n = len(args.fractional)
    labels        = args.label if len(args.label) == n else [f"Sample {i+1}" for i in range(n)]
    sample_colors = [DEFAULT_SAMPLE_COLORS[i % len(DEFAULT_SAMPLE_COLORS)] for i in range(n)]
    sample_styles = [DEFAULT_SAMPLE_STYLES[i % len(DEFAULT_SAMPLE_STYLES)] for i in range(n)]

    all_tables = [load_fractional_table(p) for p in args.fractional]
    lookups = [build_key_lookup(df) for df in all_tables]
    keys = filter_missing_keys(keys, lookups)
    if not keys:
        write_empty_plot(
            args.output,
            "No selected taxa had matching rows in fractional profile tables.",
        )
        sys.exit(0)

    pages_written = 0
    with PdfPages(args.output) as pdf:
        for key in keys:
            all_dfs = [
                subset_for_key(df, lookup, key)
                for df, lookup in zip(all_tables, lookups)
            ]

            n_bins_vals = [int(df["bin"].max()) + 1 for df in all_dfs if not df.empty]
            if not n_bins_vals:
                continue
            n_bins = max(n_bins_vals)

            all_totals, all_uncs, all_read_counts = [], [], []
            for df in all_dfs:
                tot, unc, rc = build_arrays(df, strata, n_bins)
                all_totals.append(tot)
                all_uncs.append(unc)
                all_read_counts.append(rc)

            fig = make_plot(
                all_totals, all_uncs, all_read_counts,
                labels, sample_colors, sample_styles,
                strata, n_bins, key, output_path=None,
            )
            pdf.savefig(fig, bbox_inches="tight")
            plt.close(fig)
            pages_written += 1

        if pages_written == 0:
            # Keep Snakemake outputs consistent even when requested taxa are absent.
            fig, ax = plt.subplots(figsize=(8.5, 4.5))
            ax.axis("off")
            ax.text(
                0.5, 0.62,
                "No fractional damage plots were generated",
                ha="center", va="center", fontsize=13, fontweight="bold",
            )
            ax.text(
                0.5, 0.42,
                "Selected taxa had no matching rows in the fractional profile table.",
                ha="center", va="center", fontsize=10,
            )
            ax.text(
                0.5, 0.28,
                "Try --species/--taxid, or verify species naming in --hits vs *_fractional_profile.tsv.",
                ha="center", va="center", fontsize=9, color="#555555",
            )
            pdf.savefig(fig, bbox_inches="tight")
            plt.close(fig)
            pages_written = 1

    print(f"Saved {args.output} ({pages_written} page(s))", file=sys.stderr)


if __name__ == "__main__":
    main()

File: scripts/plot_damage_fractional.R
#!/usr/bin/env Rscript
# ---------------------------------------------------------------------------
# plot_damage_fractional.R
#
# Fractional-position aDNA damage profiles from the pre-computed profiles
# written by aggregate_sample.py (*_damage_fractional_profile.tsv).
#
#   X axis : fractional position along the read (0 = 5', 1 = 3')
#   Colour : 5' half red, 3' half blue (mapDamage convention)
#   Y axis : fraction of unclassified k-mers (%)
#
# One page per taxon, strata stacked in one column with free y scales.
# Stratifying by read length separates genuine terminal damage (present in all
# strata) from the read-length artifact, where short reads have overlapping
# damage zones and no flat interior.
#
# --strata may request wider bins than the profile table stores: a requested
# stratum spanning several stored ones is summed on read (see map_strata), so
# regrouping read lengths needs no re-screening.
# ---------------------------------------------------------------------------

suppressPackageStartupMessages({
  library(ggplot2)
  library(readr)
  library(dplyr)
})

script_dir <- function() {
  a <- commandArgs(trailingOnly = FALSE)
  f <- sub("^--file=", "", a[grepl("^--file=", a)])
  if (length(f)) dirname(normalizePath(f)) else "."
}
source(file.path(script_dir(), "plot_common.R"))

# matches dataset_example/config/config.yaml; the workflow always passes --strata
DEFAULT_STRATA <- c("31-55", "56-75", "76-100")
STRATUM_COLOURS <- c("#d73027", "#fc8d59", "#4575b4", "#313695")
SAMPLE_COLOURS <- c("#d6604d", "#2166ac", "#4dac26", "#8073ac")
SAMPLE_LINETYPES <- c("solid", "dashed", "dotdash", "dotted")

SPEC <- list(
  fractional          = list(type = "character", nargs = "+", default = character(0)),
  output              = list(type = "character", nargs = 1,   default = NULL),
  label               = list(type = "character", nargs = "+", default = character(0)),
  strata              = list(type = "character", nargs = "+", default = character(0)),
  taxid               = list(type = "integer",   nargs = 1,   default = NULL),
  species             = list(type = "character", nargs = 1,   default = NULL),
  global              = list(type = "character", nargs = 1,   default = NULL),
  hits                = list(type = "character", nargs = 1,   default = NULL),
  sample_id           = list(type = "character", nargs = 1,   default = NULL),
  hits_required_flags = list(type = "character", nargs = "+", default = character(0)),
  max_keys            = list(type = "integer",   nargs = 1,   default = 200L),
  pvalue_threshold    = list(type = "double",    nargs = 1,   default = 0.05)
)

args <- parse_cli(commandArgs(trailingOnly = TRUE), SPEC)
stopifnot(length(args$fractional) > 0, !is.null(args$output))

strata <- if (length(args$strata) > 0) args$strata else DEFAULT_STRATA
n_samp <- length(args$fractional)
labels <- if (length(args$label) == n_samp) args$label else paste("Sample", seq_len(n_samp))
colours <- SAMPLE_COLOURS[(seq_len(n_samp) - 1) %% length(SAMPLE_COLOURS) + 1]
ltypes  <- SAMPLE_LINETYPES[(seq_len(n_samp) - 1) %% length(SAMPLE_LINETYPES) + 1]

# ---------------------------------------------------------------------------
# Which taxa to plot
# ---------------------------------------------------------------------------

resolve_keys_from_global <- function(path, pthr) {
  if (is.null(path)) return(character(0))
  df <- suppressWarnings(readr::read_tsv(path, col_types = readr::cols(
    .default = readr::col_guess(), species_name = readr::col_character()
  ), progress = FALSE))
  if (nrow(df) == 0L) return(character(0))
  key_col <- if ("species_name" %in% names(df)) "species_name" else "taxid"
  if ("damage_pvalue" %in% names(df)) {
    sig <- df$damage_score > 0 & df$damage_pvalue < pthr
    sig_keys <- unique(stats::na.omit(df[[key_col]][sig]))
    if (length(sig_keys) > 0) {
      cand <- df[df[[key_col]] %in% sig_keys, , drop = FALSE]
      agg <- aggregate(cand$damage_pvalue, by = list(key = cand[[key_col]]),
                       FUN = function(v) mean(v, na.rm = TRUE))
      agg <- agg[order(agg$x), , drop = FALSE]
      message("Auto-selected ", nrow(agg), " taxa with damage_pvalue < ", pthr, ".")
      return(agg$key)
    }
    message("No taxa pass p < ", pthr, ". Falling back to top 3 by score.")
  }
  agg <- aggregate(df$damage_score, by = list(key = df[[key_col]]),
                   FUN = function(v) mean(v, na.rm = TRUE))
  utils::head(agg[order(-agg$x), , drop = FALSE]$key, 3)
}

use_hits <- !is.null(args$hits) && !is.null(args$sample_id)
if (use_hits) {
  sel <- select_hit_species(args$hits, args$sample_id,
                            resolve_required_hit_flags(args$hits_required_flags),
                            args$max_keys)
  keys <- sel$species
  if (length(keys) == 0L) {
    message("WARNING: no hit species matched all required hit flags for this sample.")
  }
  if (sel$n_truncated > 0L) {
    message("WARNING: selected hit species truncated to ", length(keys),
            " (dropped ", sel$n_truncated, " by --max-keys).")
  }
} else if (!is.null(args$taxid)) {
  keys <- args$taxid
} else if (!is.null(args$species)) {
  keys <- args$species
} else {
  keys <- resolve_keys_from_global(args$global, args$pvalue_threshold)
  if (args$max_keys > 0 && length(keys) > args$max_keys) {
    message("WARNING: auto-selected taxa truncated to ", args$max_keys,
            " (dropped ", length(keys) - args$max_keys, " by --max-keys).")
    keys <- keys[seq_len(args$max_keys)]
  }
}

if (length(keys) == 0L) {
  write_empty_plot(
    args$output,
    "No fractional damage plots were generated",
    "No taxa were selected from --hits, explicit args, or --global fallback."
  )
  quit(save = "no", status = 0)
}

# ---------------------------------------------------------------------------
# Load fractional tables
# ---------------------------------------------------------------------------

wanted <- c("taxid", "species_name", "stratum", "bin", "n_reads",
            "n_total", "n_unclassified")

load_fractional <- function(path) {
  d <- suppressWarnings(readr::read_tsv(path, col_types = readr::cols(
    .default = readr::col_guess(),
    species_name = readr::col_character(),
    stratum = readr::col_character()
  ), progress = FALSE))
  d[, intersect(wanted, names(d)), drop = FALSE]
}

tables <- lapply(args$fractional, load_fractional)

key_col_for <- function(d, key) if (is.character(key)) "species_name" else "taxid"

available <- unique(unlist(lapply(tables, function(d) {
  if (is.character(keys[[1]]) && "species_name" %in% names(d)) {
    as.character(stats::na.omit(d$species_name))
  } else if ("taxid" %in% names(d)) {
    stats::na.omit(d$taxid)
  } else NULL
})))
if (length(available) > 0) {
  kept <- keys[keys %in% available]
  if (length(kept) < length(keys)) {
    message("WARNING: dropped ", length(keys) - length(kept),
            " selected taxa with no fractional rows.")
  }
  keys <- kept
}

if (length(keys) == 0L) {
  write_empty_plot(
    args$output,
    "No fractional damage plots were generated",
    "No selected taxa had matching rows in fractional profile tables."
  )
  quit(save = "no", status = 0)
}

# ---------------------------------------------------------------------------
# Per-page plot
# ---------------------------------------------------------------------------

#' Map stored strata onto requested strata.
#'
#' A requested stratum may be the union of several stored ones: the profile
#' table holds whatever strata screen_unit accumulated, and asking for a wider
#' bin (e.g. "31-55" over stored "31-40" and "41-55") is exactly a sum. Every
#' read falls in exactly one stored stratum and fractional binning depends only
#' on that read's own k-mer positions, so n_total, n_unclassified and n_reads
#' are additive across stored strata with no re-screening required.
#'
#' A stored stratum that only partially overlaps a requested one cannot be
#' split and is dropped with a warning.
map_strata <- function(stored, requested) {
  rng <- function(s) {
    parts <- strsplit(s, "-", fixed = TRUE)
    do.call(rbind, lapply(parts, function(p) as.numeric(p[1:2])))
  }
  sr <- rng(stored); rr <- rng(requested)
  out <- list()
  for (j in seq_along(requested)) {
    inside <- which(sr[, 1] >= rr[j, 1] & sr[, 2] <= rr[j, 2])
    partial <- which(sr[, 1] <= rr[j, 2] & sr[, 2] >= rr[j, 1])
    partial <- setdiff(partial, inside)
    if (length(partial) > 0L) {
      message("WARNING: stored stratum/strata ", paste(stored[partial], collapse = ", "),
              " partially overlap requested ", requested[j], "; cannot split, dropped.")
    }
    if (length(inside) == 0L) next
    if (length(inside) > 1L) {
      message("NOTE: requested stratum ", requested[j], " summed from stored ",
              paste(stored[inside], collapse = " + "), ".")
    }
    out[[length(out) + 1L]] <- data.frame(
      stored = stored[inside], requested = requested[j], stringsAsFactors = FALSE
    )
  }
  if (length(out) == 0L) return(data.frame(stored = character(0), requested = character(0)))
  dplyr::bind_rows(out)
}

build_page <- function(key) {
  rows <- list()
  reads <- list()
  for (i in seq_len(n_samp)) {
    d <- tables[[i]]
    kc <- key_col_for(d, key)
    if (!kc %in% names(d)) next
    sub <- d[d[[kc]] == key, , drop = FALSE]
    if (nrow(sub) == 0L) next

    m <- map_strata(unique(sub$stratum), strata)
    if (nrow(m) == 0L) next
    sub$requested <- m$requested[match(sub$stratum, m$stored)]
    sub <- sub[!is.na(sub$requested), , drop = FALSE]
    if (nrow(sub) == 0L) next

    n_bins <- max(sub$bin, na.rm = TRUE) + 1L

    # reads per requested stratum: n_reads is constant within a stored stratum,
    # so take it once per stored stratum and sum across the contributors
    per_stored <- sub[!duplicated(sub$stratum), c("stratum", "requested", "n_reads")]
    reads[[length(reads) + 1L]] <- stats::aggregate(
      per_stored$n_reads, by = list(requested = per_stored$requested), FUN = sum
    )

    agg <- stats::aggregate(
      cbind(n_total, n_unclassified) ~ requested + bin, data = sub, FUN = sum
    )
    agg <- agg[agg$n_total > 0, , drop = FALSE]
    if (nrow(agg) == 0L) next

    rows[[length(rows) + 1L]] <- data.frame(
      stratum  = agg$requested,
      frac     = (agg$bin + 0.5) / n_bins,
      pct      = agg$n_unclassified / agg$n_total * 100,
      sample_i = i,
      stringsAsFactors = FALSE
    )
  }
  if (length(rows) == 0L) return(NULL)
  dat <- dplyr::bind_rows(rows)
  reads <- dplyr::bind_rows(reads)

  # Read counts belong in the facet strip, not the series label: with one
  # sample and four strata a per-stratum series label produced four legend
  # entries for what is a single series.
  strip <- vapply(strata, function(s) {
    n <- sum(reads$x[reads$requested == s], na.rm = TRUE)
    if (n > 0) sprintf("%s bp  (n=%s)", sub("-", "–", s), format(n, big.mark = ","))
    else paste0(sub("-", "–", s), " bp")
  }, character(1))

  dat$stratum <- factor(dat$stratum, levels = strata, labels = strip)
  dat <- dat[!is.na(dat$stratum), , drop = FALSE]
  if (nrow(dat) == 0L) return(NULL)
  dat$series <- labels[dat$sample_i]
  dat <- dat[order(dat$stratum, dat$series, dat$frac), , drop = FALSE]

  series_levels <- unique(dat$series)
  lty <- ltypes[vapply(series_levels, function(s) dat$sample_i[match(s, dat$series)], integer(1))]
  names(lty) <- series_levels
  dat$series <- factor(dat$series, levels = series_levels)

  # Split each line at the midpoint so the 5' half is red and the 3' half blue,
  # matching plot_damage_profile.R and the mapDamage convention. Bin centres
  # never land exactly on 0.5, so a point is interpolated there and given to
  # both halves — otherwise the two segments would be drawn with a visible gap.
  dat <- do.call(rbind, lapply(split(dat, list(dat$stratum, dat$series), drop = TRUE),
    function(d) {
      d <- d[order(d$frac), , drop = FALSE]
      lo <- d[d$frac <= 0.5, , drop = FALSE]
      hi <- d[d$frac >  0.5, , drop = FALSE]
      if (nrow(lo) > 0L && nrow(hi) > 0L) {
        x1 <- lo$frac[nrow(lo)]; y1 <- lo$pct[nrow(lo)]
        x2 <- hi$frac[1];        y2 <- hi$pct[1]
        ym <- y1 + (y2 - y1) * (0.5 - x1) / (x2 - x1)
        mid_lo <- lo[nrow(lo), , drop = FALSE]; mid_lo$frac <- 0.5; mid_lo$pct <- ym
        mid_hi <- hi[1, , drop = FALSE];        mid_hi$frac <- 0.5; mid_hi$pct <- ym
        lo <- rbind(lo, mid_lo); hi <- rbind(mid_hi, hi)
      }
      if (nrow(lo)) lo$half <- "5prime"
      if (nrow(hi)) hi$half <- "3prime"
      rbind(lo, hi)
    }))
  dat$half <- factor(dat$half, levels = c("5prime", "3prime"))
  end_pal <- c(`5prime` = "#d6604d", `3prime` = "#2166ac")

  ggplot(dat, aes(x = .data$frac, y = .data$pct,
                  colour = .data$half, linetype = .data$series,
                  group = interaction(.data$series, .data$half))) +
    geom_vline(xintercept = 0.5, linetype = "dotted", colour = "grey60",
               linewidth = 0.3) +
    geom_line(linewidth = 0.7, alpha = 0.9) +
    # no colour guide: the x axis and midpoint rule already say which end is which
    scale_colour_manual(values = end_pal, guide = "none") +
    scale_linetype_manual(values = lty, name = NULL,
                          guide = if (n_samp > 1L) "legend" else "none") +
    # Stacked in one column with per-stratum y ranges, not anchored at zero.
    # Strata are independent measurements at different read lengths, so free
    # scales cost little; anchoring at zero left the curve using a fifth to a
    # third of each panel whenever the baseline unclassified rate was high.
    facet_grid(stratum ~ ., scales = "free_y", switch = "y") +
    scale_x_continuous(limits = c(0, 1), expand = expansion(mult = 0.01)) +
    scale_y_continuous(expand = expansion(mult = 0.08)) +
    labs(
      title = "aDNA damage profiles — fractional position by read length",
      subtitle = as.character(key),
      x = "Fractional position (0 = 5′, 1 = 3′)",
      y = "Unclassified k-mers (%)",
      # name the actual shortest stratum: it is configurable and may be a
      # union of stored strata, so a hardcoded "31-40 bp" goes stale
      caption = paste0(
        "Short reads (", sub("-", "\u2013", strata[1]), " bp): damage zones overlap ",
        "across the entire read, no flat interior. ",
        "Longer reads show a flat central plateau."
      )
    ) +
    theme_screen() +
    theme(legend.position = "top", legend.direction = "vertical",
          panel.spacing.y = unit(0.5, "lines"), strip.placement = "outside")
}

# ---------------------------------------------------------------------------
# Write multi-page PDF
# ---------------------------------------------------------------------------

open_pdf_device(args$output, width = 7.5,
                height = max(5, 1.9 * length(strata) + 1.6))

pages <- 0L
for (k in keys) {
  p <- build_page(k)
  if (is.null(p)) next
  print(p)
  pages <- pages + 1L
}
if (pages == 0L) {
  print(
    ggplot() +
      annotate("text", x = 0.5, y = 0.6,
               label = "No fractional damage plots were generated",
               size = 4.6, fontface = "bold") +
      annotate("text", x = 0.5, y = 0.42,
               label = "Selected taxa had no matching rows in the fractional profile table.",
               size = 3.4) +
      xlim(0, 1) + ylim(0, 1) + theme_void()
  )
  pages <- 1L
}
invisible(dev.off())
message("Saved ", args$output, " (", pages, " page(s))")
